File: backend/jasprchain/consensus/finality.py
"""Committee Finality Engine
Mirrors Rust finality module

Implements:
- Rotating committee for pre-commit + commit
- 2/3 threshold BLS aggregation
- <2s deterministic finality
"""
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Set
from datetime import datetime, timezone
import asyncio

from .block import Block, BlockHeader
from .validator import ValidatorSet, Validator
from ..crypto.bls import (
    BLSSignature, aggregate_signatures, verify_aggregate, 
    bls_sign, BLSPublicKey
)


@dataclass
class VoteMessage:
    """Vote message from a validator"""
    block_hash: str
    height: int
    validator_address: str
    signature: BLSSignature
    vote_type: str  # 'prevote' or 'precommit'
    timestamp: int


@dataclass
class CommitteeAttestation:
    """Aggregated committee attestation for a block
    
    Once 2/3 of validators sign, block is finalized
    """
    block_hash: str
    height: int
    aggregate_signature: bytes
    signers: List[str]  # Validator addresses
    voting_power: int
    total_voting_power: int
    finalized: bool = False
    finality_timestamp: Optional[int] = None
    
    @property
    def participation_rate(self) -> float:
        if self.total_voting_power == 0:
            return 0.0
        return self.voting_power / self.total_voting_power * 100
    
    def to_dict(self) -> dict:
        return {
            'block_hash': self.block_hash,
            'height': self.height,
            'aggregate_signature': self.aggregate_signature.hex() if self.aggregate_signature else '',
            'signers': self.signers,
            'voting_power': self.voting_power,
            'total_voting_power': self.total_voting_power,
            'participation_rate': round(self.participation_rate, 2),
            'finalized': self.finalized,
            'finality_timestamp': self.finality_timestamp
        }


class FinalityEngine:
    """Manages block finality through BLS attestations
    
    Implements JasprChain finality rules:
    - Safety with <1/3 Byzantine stake
    - Deterministic state root verification
    - No settlement rollback once finalized
    - <2s target finality time
    """
    
    # Target finality time in milliseconds
    TARGET_FINALITY_MS = 2000
    
    def __init__(self, validator_set: ValidatorSet):
        self.validator_set = validator_set
        
        # Vote storage by block hash
        self._prevotes: Dict[str, List[VoteMessage]] = {}
        self._precommits: Dict[str, List[VoteMessage]] = {}
        
        # Finalized blocks
        self._attestations: Dict[str, CommitteeAttestation] = {}
        
        # Timing
        self._block_proposal_times: Dict[str, int] = {}
    
    def on_block_proposed(self, block: Block):
        """Record when a block was proposed"""
        self._block_proposal_times[block.hash] = int(
            datetime.now(timezone.utc).timestamp() * 1000
        )
    
    def add_prevote(
        self, 
        block_hash: str, 
        height: int, 
        validator_address: str,
        signature: BLSSignature
    ) -> bool:
        """Add a prevote from a validator"""
        vote = VoteMessage(
            block_hash=block_hash,
            height=height,
            validator_address=validator_address,
            signature=signature,
            vote_type='prevote',
            timestamp=int(datetime.now(timezone.utc).timestamp() * 1000)
        )
        
        if block_hash not in self._prevotes:
            self._prevotes[block_hash] = []
        
        # Check for duplicate
        if any(v.validator_address == validator_address for v in self._prevotes[block_hash]):
            return False
        
        self._prevotes[block_hash].append(vote)
        return True
    
    def add_precommit(
        self, 
        block_hash: str, 
        height: int, 
        validator_address: str,
        signature: BLSSignature
    ) -> Optional[CommitteeAttestation]:
        """Add a precommit and check for finality"""
        vote = VoteMessage(
            block_hash=block_hash,
            height=height,
            validator_address=validator_address,
            signature=signature,
            vote_type='precommit',
            timestamp=int(datetime.now(timezone.utc).timestamp() * 1000)
        )
        
        if block_hash not in self._precommits:
            self._precommits[block_hash] = []
        
        # Check for duplicate
        if any(v.validator_address == validator_address for v in self._precommits[block_hash]):
            return None
        
        self._precommits[block_hash].append(vote)
        
        # Check if we have 2/3 threshold
        return self._check_finality(block_hash, height)
    
    def _check_finality(self, block_hash: str, height: int) -> Optional[CommitteeAttestation]:
        """Check if block has reached finality"""
        precommits = self._precommits.get(block_hash, [])
        
        if not precommits:
            return None
        
        # Calculate voting power
        voting_power = 0
        signers = []
        signatures = []
        
        for vote in precommits:
            validator = self.validator_set.get_validator(vote.validator_address)
            if validator and validator.active and not validator.jailed:
                voting_power += validator.voting_power
                signers.append(vote.validator_address)
                signatures.append(vote.signature)
        
        total_voting_power = self.validator_set.total_voting_power()
        
        # Check 2/3 threshold
        finalized = self.validator_set.has_supermajority(voting_power)
        
        # Aggregate signatures
        aggregate_sig = aggregate_signatures(signatures) if signatures else b''
        
        now = int(datetime.now(timezone.utc).timestamp() * 1000)
        proposal_time = self._block_proposal_times.get(block_hash, now)
        
        attestation = CommitteeAttestation(
            block_hash=block_hash,
            height=height,
            aggregate_signature=aggregate_sig,
            signers=signers,
            voting_power=voting_power,
            total_voting_power=total_voting_power,
            finalized=finalized,
            finality_timestamp=now if finalized else None
        )
        
        if finalized:
            self._attestations[block_hash] = attestation
        
        return attestation
    
    def get_attestation(self, block_hash: str) -> Optional[CommitteeAttestation]:
        """Get attestation for a block"""
        return self._attestations.get(block_hash)
    
    def is_finalized(self, block_hash: str) -> bool:
        """Check if a block is finalized"""
        attestation = self._attestations.get(block_hash)
        return attestation is not None and attestation.finalized
    
    def get_finality_time(self, block_hash: str) -> Optional[int]:
        """Get finality time in milliseconds"""
        proposal_time = self._block_proposal_times.get(block_hash)
        attestation = self._attestations.get(block_hash)
        
        if proposal_time and attestation and attestation.finality_timestamp:
            return attestation.finality_timestamp - proposal_time
        return None

File: rust-core/jasprchain/src/consensus/finality.rs
//! Committee finality engine with BLS aggregation

use crate::consensus::{Block, ValidatorSet};
use std::collections::HashMap;
use std::time::{SystemTime, UNIX_EPOCH};

/// Aggregated committee attestation for a block
pub struct CommitteeAttestation {
    pub block_hash: String,
    pub height: u64,
    pub aggregate_signature: Vec<u8>,
    pub signers: Vec<String>,
    pub voting_power: u64,
    pub total_voting_power: u64,
    pub finalized: bool,
    pub finality_timestamp: Option<u64>,
}

impl CommitteeAttestation {
    pub fn participation_rate(&self) -> f64 {
        if self.total_voting_power == 0 {
            return 0.0;
        }
        (self.voting_power as f64 / self.total_voting_power as f64) * 100.0
    }
}

/// Vote message from a validator
struct VoteMessage {
    block_hash: String,
    height: u64,
    validator_address: String,
    signature: Vec<u8>,
    timestamp: u64,
}

/// Manages block finality through BLS attestations
pub struct FinalityEngine {
    prevotes: HashMap<String, Vec<VoteMessage>>,
    precommits: HashMap<String, Vec<VoteMessage>>,
    attestations: HashMap<String, CommitteeAttestation>,
    proposal_times: HashMap<String, u64>,
}

impl FinalityEngine {
    pub const TARGET_FINALITY_MS: u64 = 2000;
    
    pub fn new() -> Self {
        Self {
            prevotes: HashMap::new(),
            precommits: HashMap::new(),
            attestations: HashMap::new(),
            proposal_times: HashMap::new(),
        }
    }
    
    pub fn on_block_proposed(&mut self, block: &Block) {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;
        self.proposal_times.insert(block.hash(), now);
    }
    
    pub fn add_prevote(
        &mut self,
        block_hash: &str,
        height: u64,
        validator: &str,
        signature: Vec<u8>,
    ) -> bool {
        let votes = self.prevotes.entry(block_hash.to_string()).or_default();
        
        if votes.iter().any(|v| v.validator_address == validator) {
            return false;
        }
        
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;
        
        votes.push(VoteMessage {
            block_hash: block_hash.to_string(),
            height,
            validator_address: validator.to_string(),
            signature,
            timestamp: now,
        });
        
        true
    }
    
    pub fn add_precommit(
        &mut self,
        block_hash: &str,
        height: u64,
        validator: &str,
        signature: Vec<u8>,
        validator_set: &ValidatorSet,
    ) -> Option<CommitteeAttestation> {
        let votes = self.precommits.entry(block_hash.to_string()).or_default();
        
        if votes.iter().any(|v| v.validator_address == validator) {
            return None;
        }
        
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;
        
        votes.push(VoteMessage {
            block_hash: block_hash.to_string(),
            height,
            validator_address: validator.to_string(),
            signature,
            timestamp: now,
        });
        
        self.check_finality(block_hash, height, validator_set)
    }
    
    fn check_finality(
        &mut self,
        block_hash: &str,
        height: u64,
        validator_set: &ValidatorSet,
    ) -> Option<CommitteeAttestation> {
        let votes = self.precommits.get(block_hash)?;
        
        let mut voting_power = 0u64;
        let mut signers = Vec::new();
        let mut signatures = Vec::new();
        
        for vote in votes {
            if let Some(v) = validator_set.get_validator(&vote.validator_address) {
                if v.active && !v.jailed {
                    voting_power += v.voting_power();
                    signers.push(vote.validator_address.clone());
                    signatures.extend(&vote.signature);
                }
            }
        }
        
        let total = validator_set.total_voting_power();
        let finalized = validator_set.has_supermajority(voting_power);
        
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;
        
        let attestation = CommitteeAttestation {
            block_hash: block_hash.to_string(),
            height,
            aggregate_signature: signatures,
            signers,
            voting_power,
            total_voting_power: total,
            finalized,
            finality_timestamp: if finalized { Some(now) } else { None },
        };
        
        if finalized {
            self.attestations.insert(block_hash.to_string(), attestation.clone());
        }
        
        Some(attestation)
    }
    
    pub fn is_finalized(&self, block_hash: &str) -> bool {
        self.attestations
            .get(block_hash)
            .map(|a| a.finalized)
            .unwrap_or(false)
    }
    
    pub fn get_finality_time(&self, block_hash: &str) -> Option<u64> {
        let proposal_time = self.proposal_times.get(block_hash)?;
        let attestation = self.attestations.get(block_hash)?;
        attestation.finality_timestamp.map(|t| t - proposal_time)
    }
}

impl Default for FinalityEngine {
    fn default() -> Self {
        Self::new()
    }
}
